mithril-aggregator/src/file_uploaders/gcp_uploader.rs

```diff
@@ -5,93 +5,243 @@ use cloud_storage::{
+
+        Ok(Self {
             bucket,
             use_cdn_domain,
             logger: logger.new_with_component_name::<Self>(),
-        }
+            client: Client::default(),
+        })
     }
 
-    fn get_location(&self, filename: &str) -> FileUri {
+    fn get_location(&self, remote_file_path: &CloudRemotePath) -> FileUri {
         let mut uri = vec![];
         if !self.use_cdn_domain {
             uri.push("storage.googleapis.com");
         }
         uri.push(&self.bucket);
-        uri.push(filename);
+        let file_path = remote_file_path.to_string();
+        uri.push(&file_path);
 
         FileUri(format!("https://{}", uri.join("/")))
     }
 }
 
 #[async_trait]
-impl FileUploader for GcpUploader {
-    async fn upload(&self, filepath: &Path) -> StdResult<FileUri> {
-        if env::var("GOOGLE_APPLICATION_CREDENTIALS_JSON").is_err() {
-            return Err(anyhow!(
-                "Missing GOOGLE_APPLICATION_CREDENTIALS_JSON environment variable".to_string()
-            ));
+impl CloudBackendUploader for GcpBackendUploader {
+    async fn file_exists(&self, remote_file_path: &CloudRemotePath) -> StdResult<Option<FileUri>> {
+        info!(self.logger, "Reading file metadata {remote_file_path}");
+        let file_uri = match self
+            .client
+            .object()
+            .read(&self.bucket, &remote_file_path.to_string())
+            .await
+            .with_context(|| "remote reading file metadata failure")
+        {
+            Ok(_) => {
+                info!(self.logger, "Found file metadata {remote_file_path}");
+
+                Some(self.get_location(remote_file_path))
+            }
+            Err(_) => {
+                info!(self.logger, "Missing file metadata {remote_file_path}");
+
+                None
+            }
         };
 
-        let filename = filepath.file_name().unwrap().to_str().unwrap();
+        Ok(file_uri)
+    }
 
-        info!(self.logger, "Uploading {filename}");
-        let client = Client::default();
-        let file = tokio::fs::File::open(filepath).await.unwrap();
+    async fn upload_file(
+        &self,
+        local_file_path: &Path,
+        remote_file_path: &CloudRemotePath,
+    ) -> StdResult<FileUri> {
+        info!(
+            self.logger,
+            "Uploading {local_file_path:?} to {remote_file_path}"
+        );
+        let file = tokio::fs::File::open(local_file_path).await.unwrap();
         let stream = FramedRead::new(file, BytesCodec::new());
-        client
+        self.client
             .object()
             .create_streamed(
                 &self.bucket,
                 stream,
                 None,
-                filename,
+                &remote_file_path.to_string(),
                 "application/octet-stream",
             )
             .await
             .with_context(|| "remote uploading failure")?;
+        info!(
+            self.logger,
+            "Uploaded {local_file_path:?} to {remote_file_path}"
```

Comment:
You should use the `display` method when logging `Path`
```suggestion
            "Uploaded {} to {remote_file_path}", local_file_path.display()
```

Reply:
Done